Jenkinsfile

```diff
@@ -1,62 +1,89 @@
-      if (!dockerImageExists(docker_image)) {
-            sh "echo 'ENV GIT_SHA ${GIT_COMMIT}' >> Dockerfile"
-
-            // Docker always copies data as root. Chowning files to root before
-            // the build command will prevent docker cache invalidation due to
-            // different file metadata. When we upgrade to a newer docker
-            // version we should probably use --chown flag for COPY. See
-            // https://github.com/moby/moby/pull/34263
-            dockerRun('debian', [
-                    "docker_args": "-v `pwd`:/app",
-                    "cmd": "cd /app && chown -c root.root -R requirements bower.json package.json && chmod -c a+rw -R requirements bower.json package.json",
-                    "bash_wrap": true
-                ])
-            dockerImagePull( "${config.project.docker_name}:latest")
-            dockerImageBuild(docker_image, [
-                    "pull": true,
-                    "cache_from": "${config.project.docker_name}:latest",
-                ])
-      }
-      else {
-        echo "Image ${docker_image} already exists."
-      }
-    }
+            try {
+                config = readYaml file: "jenkins.yml"
+            }
+            catch (e) {
+                config = []
+            }
+            println "config ==> ${config}"
 
-    stage("Upload Images") {
-      dockerImagePush(docker_image, "mozjenkins-docker-hub")
-      onBranch("master") {
-        dockerImageTag(docker_image, "${config.project.docker_name}:latest")
-        dockerImagePush("${config.project.docker_name}:latest", "mozjenkins-docker-hub")
-      }
-    }
-  }
+            if (!config || (config && config.pipeline && config.pipeline.enabled == false)) {
+                println "Pipeline disabled."
+            }
+        }
+
+        docker_image = "${config.project.docker_name}:full-${GIT_COMMIT_SHORT}"
+
+        stage("Build") {
+            if (!dockerImageExists(docker_image)) {
+                sh "GIT_SHA=${GIT_COMMIT_SHORT} LOCALE_ENV=production ./docker/bin/build-docker-images.sh"
+            }
+            else {
+                echo "Image ${docker_image} already exists."
+            }
+        }
+
+        stage("Upload Images") {
+            dockerImagePush("${config.project.docker_name}:full-${GIT_COMMIT_SHORT}", "mozjenkins-docker-hub")
+            dockerImagePush("${config.project.docker_name}:full-no-locales-${GIT_COMMIT_SHORT}", "mozjenkins-docker-hub")
+            dockerImagePush("${config.project.docker_name}:locales-${GIT_COMMIT_SHORT}", "mozjenkins-docker-hub")
+            dockerImagePush("${config.project.docker_name}:staticfiles-${GIT_COMMIT_SHORT}", "mozjenkins-docker-hub")
+            dockerImagePush("${config.project.docker_name}:base-dev-${GIT_COMMIT_SHORT}", "mozjenkins-docker-hub")
+            dockerImagePush("${config.project.docker_name}:base-${GIT_COMMIT_SHORT}", "mozjenkins-docker-hub")
+        }
 
-  milestone()
+        stage("Run Flake8") {
```

Comment:
To me it makes more sense to run these tests before pushing the images to docker hub? Is there a reason to push images that fail tests?

Reply:
Yes and no.

Ideally I agree we shouldn't be pushing images before tests pass. And that's why we don't mark images as `latest` unless tests pass.

In CircleCI we got this annoying `mysql has gone away` error msg and I changed the code to push images before running the tests anyway, so I could get an image and test locally or push directly if it's about a change that didn't require running the tests.

The problem with Kitsune tests is that they take toooo long so we need to make steps towards improving the speed (`test --parallel` with Django 1.11 or fancier things with `py.test`) so pushing early makes things available  to start the manual work  earlier.

Reply:
On second though:

- This pipeline runs on commits pushed to `mozilla/kitsune` so why not have images pushed to docker hub independently of the status of the tests? It shouldn't hurt and it's possible helpful with debugging?

Reply:
That seems fair. Proceed!